File: analyze_mat_questions.py
#!/usr/bin/env python3
"""Analyze MAT LGS JSON files: parse failures, short_item (exact + heuristic)."""

import json
import os

MAT_DIR = "app/src/main/assets/lgs_import/mat"

PROBLEM_KEYWORDS = [
    "problemi", "problem", "oran", "yüzde", "grafik", "tablo",
    "şekilde", "aşağıdaki", "metne göre", "parçaya göre"
]


def get_stem(q):
    stem = (q.get("stem") or q.get("questionText") or "").strip()
    return stem


def get_options_raw(q):
    opts = q.get("options") or q.get("choices")
    return opts


def parse_lgs_question(o, index):
    """Replicate QuestionPackImporter.parseLgsQuestion logic.
    Returns (None, reason) on failure, or (entity_data, None) on success.
    """
    stem = (o.get("stem") or o.get("questionText") or "").strip()
    if not stem:
        return None, "stem_blank"

    options_raw = o.get("options") or o.get("choices")
    if options_raw is None:
        return None, "options_missing"

    # Filter blank options (Kotlin: optString/opt().toString, then filter isNotBlank)
    raw_opts = []
    for x in options_raw:
        s = str(x).strip() if x is not None else ""
        if s:
            raw_opts.append(s)

    if len(raw_opts) < 2:
        return None, f"fewer_than_2_options_after_filter ({len(raw_opts)} valid)"

    return {"stem": stem, "questionType": o.get("questionType", ""), "index": index}, None


def has_problem_keyword(stem):
    s = stem.lower()
    for kw in PROBLEM_KEYWORDS:
        if kw in s:
            return True
    return False


def is_short_item_heuristic(stem, question_type):
    """scan_short_items.py heuristic."""
    stem_len = len(stem)
    if question_type == "short_item":
        return True
    if stem_len < 80:
        return True
    if stem_len < 120 and not has_problem_keyword(stem):
        return True
    return False


def main():
    base = os.path.dirname(os.path.abspath(__file__))
    mat_path = os.path.join(base, MAT_DIR)

    parse_failures = []
    short_item_exact = []
    short_item_heuristic = []

    json_files = sorted([f for f in os.listdir(mat_path) if f.endswith(".json")])

    for fn in json_files:
        filepath = os.path.join(mat_path, fn)
        rel_path = os.path.join(MAT_DIR, fn)
        try:
            with open(filepath, "r", encoding="utf-8") as f:
                data = json.load(f)
        except Exception as e:
            parse_failures.append({
                "file": rel_path,
                "index": -1,
                "stem_preview": "",
                "reason": f"file_load_error: {e}"
            })
            continue

        questions = data.get("questions") or []
        for i, q in enumerate(questions):
            entity, reason = parse_lgs_question(q, i)
            if entity is None:
                stem_preview = (get_stem(q) or "(no stem)")[:80]
                parse_failures.append({
                    "file": rel_path,
                    "index": i,
                    "stem_preview": stem_preview,
                    "reason": reason
                })
            else:
                stem = entity["stem"]
                qt = entity.get("questionType") or ""

                full_q = {**q, "file": rel_path, "index": i}

                if qt == "short_item":
                    short_item_exact.append(full_q)

                if is_short_item_heuristic(stem, qt):
                    short_item_heuristic.append({
                        "file": rel_path,
                        "index": i,
                        "stem": stem,
                        "stem_length": len(stem),
                        "questionType": qt,
                        "full_question": full_q
                    })

    # Report
    print("=" * 60)
    print("1. PARSE-FAILING QUESTIONS (parseLgsQuestion returns null)")
    print("=" * 60)
    for p in parse_failures:
        print(f"  File: {p['file']}")
        print(f"  Index: {p['index']}")
        print(f"  Stem preview: {p['stem_preview'][:80]}...")
        print(f"  Reason: {p['reason']}")
        print()

    print("=" * 60)
    print("2. SHORT_ITEM (questionType == 'short_item' exact match)")
    print("=" * 60)
    for q in short_item_exact:
        stem = get_stem(q)
        print(f"  File: {q['file']}, Index: {q['index']}")
        print(f"  Stem: {stem[:100]}{'...' if len(stem) > 100 else ''}")
        print(f"  Full question data: {json.dumps(q, ensure_ascii=False, indent=2)[:500]}...")
        print()

    print("=" * 60)
    print("3. SHORT_ITEM (scan_short_items heuristic: stem<80 OR stem<120+no problem keywords OR questionType short_item)")
    print("=" * 60)
    for s in short_item_heuristic:
        print(f"  File: {s['file']}, Index: {s['index']}")
        print(f"  Stem length: {s['stem_length']}, questionType: {s['questionType']}")
        print(f"  Stem: {s['stem'][:100]}{'...' if len(s['stem']) > 100 else ''}")
        print()

    # JSON output for programmatic use
    output = {
        "parse_failures": parse_failures,
        "short_item_exact": short_item_exact,
        "short_item_heuristic": short_item_heuristic,
    }
    out_path = os.path.join(base, "mat_analysis_output.json")
    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)
    print(f"\nFull output saved to: {out_path}")
    print(f"\nSummary: {len(parse_failures)} parse failures, {len(short_item_exact)} short_item exact, {len(short_item_heuristic)} short_item heuristic")


if __name__ == "__main__":
    main()

File: analyze_mat_questions.mjs
#!/usr/bin/env node
/** Analyze MAT LGS JSON files: parse failures, short_item (exact + heuristic). */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const MAT_DIR = path.join(__dirname, "app/src/main/assets/lgs_import/mat");

const PROBLEM_KEYWORDS = [
  "problemi", "problem", "oran", "yüzde", "grafik", "tablo",
  "şekilde", "aşağıdaki", "metne göre", "parçaya göre"
];

function getStem(q) {
  return ((q.stem || q.questionText || "") + "").trim();
}

function parseLgsQuestion(o, index) {
  const stem = ((o.stem || o.questionText || "") + "").trim();
  if (!stem) return { fail: true, reason: "stem_blank" };

  const optionsRaw = o.options || o.choices;
  if (optionsRaw == null) return { fail: true, reason: "options_missing" };

  const arr = Array.isArray(optionsRaw) ? optionsRaw : [];
  const rawOpts = arr
    .map((x) => (x != null ? String(x).trim() : ""))
    .filter((s) => s.length > 0);

  if (rawOpts.length < 2) {
    return { fail: true, reason: `fewer_than_2_options_after_filter (${rawOpts.length} valid)` };
  }

  return { fail: false, stem, questionType: (o.questionType || "").trim() };
}

function hasProblemKeyword(stem) {
  const s = stem.toLowerCase();
  return PROBLEM_KEYWORDS.some((kw) => s.includes(kw));
}

function isShortItemHeuristic(stem, questionType) {
  const stemLen = stem.length;
  if (questionType === "short_item") return true;
  if (stemLen < 80) return true;
  if (stemLen < 120 && !hasProblemKeyword(stem)) return true;
  return false;
}

const parseFailures = [];
const shortItemExact = [];
const shortItemHeuristic = [];

const files = fs.readdirSync(MAT_DIR).filter((f) => f.endsWith(".json")).sort();

for (const fn of files) {
  const filepath = path.join(MAT_DIR, fn);
  const relPath = path.join("app/src/main/assets/lgs_import/mat", fn);
  let data;
  try {
    data = JSON.parse(fs.readFileSync(filepath, "utf8"));
  } catch (e) {
    parseFailures.push({
      file: relPath,
      index: -1,
      stem_preview: "",
      reason: `file_load_error: ${e.message}`,
    });
    continue;
  }

  const questions = data.questions || [];
  for (let i = 0; i < questions.length; i++) {
    const q = questions[i];
    const result = parseLgsQuestion(q, i);

    if (result.fail) {
      const stemPreview = (getStem(q) || "(no stem)").slice(0, 80);
      parseFailures.push({
        file: relPath,
        index: i,
        stem_preview: stemPreview,
        reason: result.reason,
      });
    } else {
      const fullQ = { ...q, file: relPath, index: i };
      if (result.questionType === "short_item") {
        shortItemExact.push(fullQ);
      }
      if (isShortItemHeuristic(result.stem, result.questionType)) {
        shortItemHeuristic.push({
          file: relPath,
          index: i,
          stem: result.stem,
          stem_length: result.stem.length,
          questionType: result.questionType,
          full_question: fullQ,
        });
      }
    }
  }
}

// Output
console.log("=".repeat(60));
console.log("1. PARSE-FAILING QUESTIONS (parseLgsQuestion returns null)");
console.log("=".repeat(60));
for (const p of parseFailures) {
  console.log(`  File: ${p.file}`);
  console.log(`  Index: ${p.index}`);
  console.log(`  Stem preview: ${(p.stem_preview || "").slice(0, 80)}`);
  console.log(`  Reason: ${p.reason}`);
  console.log();
}

console.log("=".repeat(60));
console.log("2. SHORT_ITEM (questionType == 'short_item' exact match)");
console.log("=".repeat(60));
for (const q of shortItemExact) {
  const stem = getStem(q);
  console.log(`  File: ${q.file}, Index: ${q.index}`);
  console.log(`  Stem: ${stem.slice(0, 100)}${stem.length > 100 ? "..." : ""}`);
  console.log(`  Full question:`, JSON.stringify(q, null, 2).slice(0, 500) + "...");
  console.log();
}

console.log("=".repeat(60));
console.log("3. SHORT_ITEM (scan_short_items heuristic)");
console.log("=".repeat(60));
for (const s of shortItemHeuristic) {
  console.log(`  File: ${s.file}, Index: ${s.index}`);
  console.log(`  Stem length: ${s.stem_length}, questionType: ${s.questionType}`);
  console.log(`  Stem: ${s.stem.slice(0, 100)}${s.stem.length > 100 ? "..." : ""}`);
  console.log();
}

const output = {
  parse_failures: parseFailures,
  short_item_exact: shortItemExact,
  short_item_heuristic: shortItemHeuristic,
};
const outPath = path.join(__dirname, "mat_analysis_output.json");
fs.writeFileSync(outPath, JSON.stringify(output, null, 2), "utf8");
console.log(`\nFull output saved to: ${outPath}`);
console.log(
  `Summary: ${parseFailures.length} parse failures, ${shortItemExact.length} short_item exact, ${shortItemHeuristic.length} short_item heuristic`
);
